### PR TITLE
fix: Improve async event loop handling in component tools, this Improves the Agent as Tool Functionality

### PR DESCRIPTION
**Pull request by edwinjosechittilappilly**
Fix: Improve async event loop handling in component tools

## Description
Enhanced the `_build_output_function` in component tools to properly manage async event loops and added comprehensive tests for agent tool functionality. This addresses issues with event loop closure and resource cleanup in async operations.

## Changes
- Added proper event loop management in `_build_output_function`
- Improved documentation with comprehensive docstrings
- Added cleanup of pending tasks to prevent resource leaks
- Added pytest cases to verify agent tool functionality

---------
This pull request introduces several important changes to improve the handling of synchronous and asynchronous methods, enhance error handling, and add new test cases. Key changes include the addition of a utility function to check if a method is asynchronous, updates to the `AgentComponent` class to handle missing chat history gracefully, and new tests for the `AgentComponent` class.

### Enhancements to method handling:

* Added `is_async_callable` utility function to check if a method is asynchronous in `src/backend/base/langflow/base/tools/component_tool.py`.
* Updated `_build_output_function` to create wrappers for both synchronous and asynchronous methods, ensuring proper event loop management for async methods in `src/backend/base/langflow/base/tools/component_tool.py`.

### Error handling improvements:

* Enhanced `AgentComponent` to handle missing chat history gracefully by adding a try-except block in `src/backend/base/langflow/components/agents/agent.py`.

### Test additions:

* Added a new test case `test_agent_as_tool` to verify the functionality of using an agent as a tool in `src/backend/tests/unit/components/agents/test_agent_component.py`.

### Minor improvements:

* Added required inputs to the `outputs` attribute of `AgentComponent` in `src/backend/base/langflow/components/agents/agent.py`.
* Updated `_set_output_required_inputs` to skip setting required inputs if they are already defined in `src/backend/base/langflow/custom/custom_component/component.py`.

### CODE REVIEW

**Comment by edwinjosechittilappilly** [comment]
Note: Merge only after 
https://github.com/langflow-ai/langflow/pull/4389 Merges to Main

**Comment by edwinjosechittilappilly** [comment]
Agent As Tool Demo
https://github.com/user-attachments/assets/25c2387f-a22d-4599-ad6a-8cc394e66f3f

File: src/backend/base/langflow/base/tools/component_tool.py
Comment (ogabrielluiz) [inline src/backend/base/langflow/base/tools/component_tool.py:1]: Hey, @edwinjosechittilappilly 

I opened #4402 (PTAL) that has implementation for this. LangChain already takes care of calling coroutines safely so we don't have to manage event loops in here.

File: src/backend/base/langflow/components/agents/agent.py
Comment (ogabrielluiz) [inline src/backend/base/langflow/components/agents/agent.py:1]: #4402 also adds the `tool_mode` param to the fields so we can define which input gets passed by the agent.

**Review by ogabrielluiz — CHANGES_REQUESTED** [review]

File: src/backend/base/langflow/base/tools/component_tool.py
Comment (edwinjosechittilappilly) [reply src/backend/base/langflow/base/tools/component_tool.py:1]: Sounds Good!

**Pull request closed without merging**